### DISCUSSION TITLE
Suggestion: reduce randomness

### DISCUSSION THREAD

**Discussion by jacksondc** [yasserfarouk/scml#79] [Ideas]
There are many random parameters. The more randomness there is in the simulation, the harder it is to learn. Maybe we can eliminate some randomness to make it more tractable. For instance, is it necessary to choose a random disposal cost before the game, and then vary that disposal cost throughout the game?

**Comment by yasserfarouk** [comment]
The parameters of the game are controlled by keyword arguments of the `generate()` method in `SCMLOneShotWorld`, `SCML2020World` and `SCML2021World` classes. It is possible to fix any parameter to one value. 

Let's take the disposal cost example. This is controlled through the following two arguments of the `generate` method:

```python
        disposal_cost: Union[np.ndarray, Tuple[float, float], float] = (0.0, 0.2),
        disposal_cost_dev: Union[np.ndarray, Tuple[float, float], float] = (0.0, 0.02),
```

The type of `disposal_cost` is common for parameters of `generate()` so I will spend a little time describing it. It is a union of three possibilities:

1. A `numpy` array which gives the `mean` disposal cost per agent **NOT SUPPORTED YET**
2. A tuple of two values giving the minimum and maximum of a uniform distribution from which we sample the mean disposal cost (that is the first step you spoke about)
3. A value that can be fixed as the mean disposal cost for all agents.

The same is true for `disposal_cost_dev` which controls the standard deviation of the distribution used per agent. The standard deviation of the distribution used during the game is whatever comes from this value multiplied by the mean.

**To fix the disposal cost for all steps, you just need to pass `disposal_cost_dev=0` to `generate()`**

The same is true for other parameters. They can be fixed by controlling what passes to the `generate()` method. If you are using the tournament interface, you can still do the same because all tournament generation methods receive `kwargs` that are passed directly to generate.

One limitation  we have with the current generate function is that you cannot fix a different disposal cost (for example) for each agent (or for each step). If you think this is useful, let me know and we will put it in our todos to extend `generate()` to support this possibility as well. This can be done by allowing you to pass a list/numpy array of exactly `n_agents` numbers/Tuples to control the value for each agent independently. For now, you can do that by simply inheriting from the World class, and overriding the values in the profiles generated by the generate method. SOmething like:

```python
class MyWorld(SCMLOneShotWOrld):
     def generate(**kwargs):
          d = super().generate(**kwargs):
          for profile in profiles:
              profile.disposal_cost_mean = ...
              profile.disposal_cost_dev = ...
```

You can now use `MyWorld` instead of `SCMLOneShotWorld` and everything should work as you wanted.

**Reply by jacksondc** [reply]
Thanks, Yasser! At the moment I can't think of a use we'd have for fixing different disposal costs for each agent, but maybe others would find it useful.
